routing/payment_lifecycle.go

```diff
@@ -2,346 +2,592 @@ package routing
+				log.Debugf("Marking payment %v permanently "+
+					"failed with no route: %v",
+					p.paymentHash, failureCode)
+
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
+
+				continue
+			}
 
-			err = p.failAttempt(result.Error)
-			if err != nil {
-				return [32]byte{}, nil, err
+			// We still have active shards, we'll wait for an
+			// outcome to be available before retrying.
+			select {
+			case err := <-shardErrors:
+				if err != nil {
+					return [32]byte{}, nil, err
+				}
+
+			case <-p.router.quit:
+				return [32]byte{}, nil, ErrRouterShuttingDown
 			}
 
+			continue
+		}
+
+		// We found a route to try, launch a new shard.
+		attempt, outcome, err := shardHandler.launchShard(rt)
+		if err != nil {
+			return [32]byte{}, nil, err
+		}
+
+		// We ew encountered a non-critical error when launching the
+		// shard, handle it
+		if outcome.err != nil {
+			log.Warnf("Failed to launch shard %v for "+
+				"payment %v: %v", attempt.AttemptID,
+				p.paymentHash, outcome.err)
+
 			// We must inspect the error to know whether it was
 			// critical or not, to decide whether we should
 			// continue trying.
-			if err := p.handleSendError(result.Error); err != nil {
+			err := shardHandler.handleSendError(
+				attempt, outcome.err,
+			)
+			if err != nil {
 				return [32]byte{}, nil, err
 			}
 
-			// Error was handled successfully, reset the attempt to
-			// indicate we want to make a new attempt.
-			p.attempt = nil
+			// Error was handled successfully, continue to make a
+			// new attempt.
 			continue
 		}
 
-		// We successfully got a payment result back from the switch.
-		log.Debugf("Payment %x succeeded with pid=%v",
-			p.payment.PaymentHash, p.attempt.AttemptID)
-
-		// Report success to mission control.
-		err = p.router.cfg.MissionControl.ReportPaymentSuccess(
-			p.attempt.AttemptID, &p.attempt.Route,
+		// Now that the shard was successfully sent, launch a go
+		// routine that will handle its result when its back.
+		shardHandler.collectResultAsync(
+			attempt, shardErrors,
 		)
+	}
+}
+
+// shardHandler holds what is necessary to send and collect the result of
+// shards.
+type shardHandler struct {
+	paymentHash lntypes.Hash
+	router      *ChannelRouter
+
+	// quit is closed to signal the sub goroutines of the payment lifecycle
+	// to stop.
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+func (p *shardHandler) Stop() {
+	close(p.quit)
+	p.wg.Wait()
+}
+
+// launchOutcome is a type returned from launchShard that indicates whether the
+// shard was successfully send onto the network.
+type launchOutcome struct {
+	// err is non-nil if a non-critical error was encountered when trying
+	// to send the shard, and we successfully updated the control tower to
+	// reflect this error. This can be errors like not enough local
+	// balance for the given route etc.
+	err error
+}
```

Comment:
Is this struct needed?

Reply:
I like it, since it more clearly distinguish the return value from the critical error case.

Reply:
This doesn't make sense to me. Can't it be a type alias then?